File: solutions/prospective/algorithms/CLRS/js/quick_sort_opt_hoare.js
/**
 * Quicksort (Optimized Hoare)
 * 
 * Improvements:
 * 1. Median-of-Three pivot (fixes O(n^2) on sorted arrays).
 * 2. Uses Hoare partition (3x fewer swaps than Lomuto).
 */
// === MEMO START ===
function medianOfThree(arr, low, high) {
    const mid = Math.floor((low + high) / 2);
    if (arr[low] > arr[mid]) [arr[low], arr[mid]] = [arr[mid], arr[low]];
    if (arr[low] > arr[high]) [arr[low], arr[high]] = [arr[high], arr[low]];
    if (arr[mid] > arr[high]) [arr[mid], arr[high]] = [arr[high], arr[mid]];

    // Median is now at mid; swap with low to use as pivot
    [arr[mid], arr[low]] = [arr[low], arr[mid]];
    return arr[low];
}

function partition(arr, low, high) {
    const pivot = medianOfThree(arr, low, high);
    let i = low - 1;
    let j = high + 1;

    while (true) {
        do {
            i++;
        } while (arr[i] < pivot);

        do {
            j--;
        } while (arr[j] > pivot);

        if (i >= j) {
            return j;
        }

        [arr[i], arr[j]] = [arr[j], arr[i]];
    }
}

function quickSort(arr, low, high) {
    if (low < high) {
        const p = partition(arr, low, high);
        quickSort(arr, low, p);
        quickSort(arr, p + 1, high);
    }
}

File: solutions/prospective/algorithms/CLRS/lua/quick_sort_opt_hoare.lua
--[[
Quicksort (Optimized Hoare)

Improvements:
1. Median-of-Three pivot (fixes O(n^2) on sorted arrays).
2. Uses Hoare partition (3x fewer swaps than Lomuto).
--]]
-- === MEMO START ===
function median_of_three(arr, low, high)
    local mid = math.floor((low + high) / 2)
    if arr[low] > arr[mid] then arr[low], arr[mid] = arr[mid], arr[low] end
    if arr[low] > arr[high] then arr[low], arr[high] = arr[high], arr[low] end
    if arr[mid] > arr[high] then arr[mid], arr[high] = arr[high], arr[mid] end
    
    -- Median is now at mid; swap with low to use as pivot
    arr[mid], arr[low] = arr[low], arr[mid]
    return arr[low]
end

function partition(arr, low, high)
    local pivot = median_of_three(arr, low, high)
    local i = low - 1
    local j = high + 1
    
    while true do
        repeat
            i = i + 1
        until arr[i] >= pivot
        
        repeat
            j = j - 1
        until arr[j] <= pivot
        
        if i >= j then
            return j
        end
        
        arr[i], arr[j] = arr[j], arr[i]
    end
end

function quick_sort(arr, low, high)
    if low < high then
        local p = partition(arr, low, high)
        quick_sort(arr, low, p)
        quick_sort(arr, p + 1, high)
    end
end

